cumulus/parachains/runtimes/assets/common/src/erc20_transactor.rs

```diff
@@ -0,0 +1,220 @@
+		AccountIdConverter: ConvertLocation<AccountId>,
+		Matcher: MatchesFungibles<H160, u128>,
+		GasLimit: Get<Weight>,
+		StorageDepositLimit: Get<BalanceOf<T>>,
+		TransfersCheckingAccount: Get<AccountId>,
+	> TransactAsset
+	for ERC20Transactor<
+		T,
+		Matcher,
+		AccountIdConverter,
+		GasLimit,
+		StorageDepositLimit,
+		AccountId,
+		TransfersCheckingAccount,
+	>
+where
+	BalanceOf<T>: Into<U256> + TryFrom<U256>,
+	MomentOf<T>: Into<U256>,
+	T::Hash: frame_support::traits::IsType<H256>,
+{
+	fn can_check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_in(_origin: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn can_check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) -> XcmResult {
+		// We don't support teleports.
+		Err(XcmError::Unimplemented)
+	}
+
+	fn check_out(_destination: &Location, _what: &Asset, _context: &XcmContext) {
+		// We don't support teleports.
+	}
+
+	fn withdraw_asset_with_surplus(
+		what: &Asset,
+		who: &Location,
+		_context: Option<&XcmContext>,
+	) -> Result<(AssetsInHolding, Weight), XcmError> {
+		tracing::trace!(
+			target: "xcm::transactor::erc20::withdraw",
+			?what, ?who,
+		);
+		let (asset_id, amount) = Matcher::matches_fungibles(what)?;
+		let who = AccountIdConverter::convert_location(who)
+			.ok_or(MatchError::AccountIdConversionFailed)?;
+		// We need to map the 32 byte checking account to a 20 byte account.
+		let checking_account_eth = T::AddressMapper::to_address(&TransfersCheckingAccount::get());
+		let checking_address = Address::from(Into::<[u8; 20]>::into(checking_account_eth));
+		let gas_limit = GasLimit::get();
+		// To withdraw, we actually transfer to the checking account.
+		// We do this using the solidity ERC20 interface.
+		let data =
+			IERC20::transferCall { to: checking_address, value: EU256::from(amount) }.abi_encode();
+		let ContractResult { result, gas_consumed, storage_deposit, .. } =
+			pallet_revive::Pallet::<T>::bare_call(
+				T::RuntimeOrigin::signed(who.clone()),
+				asset_id,
+				BalanceOf::<T>::zero(),
+				gas_limit,
+				DepositLimit::Balance(StorageDepositLimit::get()),
+				data,
+			);
+		// We need to return this surplus for the executor to allow refunding it.
+		let surplus = gas_limit.saturating_sub(gas_consumed);
+		tracing::trace!(target: "xcm::transactor::erc20::withdraw", ?gas_consumed, ?surplus, ?storage_deposit);
+		if let Ok(return_value) = result {
+			tracing::trace!(target: "xcm::transactor::erc20::withdraw", ?return_value, "Return value by withdraw_asset");
+			if return_value.did_revert() {
+				tracing::debug!(target: "xcm::transactor::erc20::withdraw", "ERC20 contract reverted");
+				Err(XcmError::FailedToTransactAsset("ERC20 contract reverted"))
+			} else {
+				let is_success = bool::abi_decode(&return_value.data, false).map_err(|error| {
```

Comment:
Was is a conscious decision to disable validation? All other users of alloy-core inside the runtime (snowbridge, pallet-revive) do enable it.

```suggestion
				let is_success = bool::abi_decode(&return_value.data, true).map_err(|error| {
```

Reply:
It's better to validate. Changed it. Also bumped `alloy-core` as per PG's suggestion so I used `abi_decode_validate` everywhere